test/huffman_codec.cpp

```diff
@@ -0,0 +1,220 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Contains utils for reading, writing and debug printing bit streams.
+
+#include <map>
+#include <sstream>
+#include <string>
+#include <unordered_map>
+
+#include "util/huffman_codec.h"
+#include "util/bit_stream.h"
+#include "gmock/gmock.h"
+
+namespace {
+
+using spvutils::HuffmanCodec;
+using spvutils::BitsToStream;
+
+const std::map<std::string, uint32_t>& GetTestSet() {
+  static const std::map<std::string, uint32_t> hist = {
+    {"a", 4},
+    {"e", 7},
+    {"f", 3},
+    {"h", 2},
+    {"i", 3},
+    {"m", 2},
+    {"n", 2},
+    {"s", 2},
+    {"t", 2},
+    {"l", 1},
+    {"o", 2},
+    {"p", 1},
+    {"r", 1},
+    {"u", 1},
+    {"x", 1},
+  };
+
+  return hist;
+}
+
+class TestBitReader {
+ public:
+  TestBitReader(const std::string& bits) : bits_(bits) {}
+
+  bool ReadBit(bool* bit) {
+    if (pos_ < bits_.length()) {
+      *bit = bits_[pos_++] == '0' ? false : true;
+      return true;
+    }
+    return false;
+  }
+
+ private:
+  std::string bits_;
+  size_t pos_ = 0;
+};
+
+TEST(Huffman, PrintTree) {
+  HuffmanCodec<std::string> huffman(GetTestSet());
+  std::stringstream ss;
+  huffman.PrintTree(ss);
+
+  const std::string expected = std::string(R"(
```

Comment:
Elegant test cases.  I'm glad you put this tree output here first!   I'm sure you're glad too.  :-)